crates/storage/provider/src/providers/database/mod.rs

```diff
@@ -0,0 +1,324 @@
+use crate::{
+    providers::state::{historical::HistoricalStateProvider, latest::LatestStateProvider},
+    traits::{BlockSource, ReceiptProvider},
+    BlockHashProvider, BlockNumProvider, BlockProvider, EvmEnvProvider, HeaderProvider,
+    ProviderError, StateProviderBox, TransactionsProvider, WithdrawalsProvider,
+};
+use reth_db::{database::Database, tables, transaction::DbTx};
+use reth_interfaces::Result;
+use reth_primitives::{
+    Block, BlockHash, BlockHashOrNumber, BlockNumber, ChainInfo, ChainSpec, Header, Receipt,
+    SealedBlock, SealedHeader, TransactionMeta, TransactionSigned, TxHash, TxNumber, Withdrawal,
+    H256, U256,
+};
+use reth_revm_primitives::primitives::{BlockEnv, CfgEnv};
+use std::{ops::RangeBounds, sync::Arc};
+use tracing::trace;
+
+mod provider;
+use provider::*;
+
+/// A common provider that fetches data from a database.
+///
+/// This provider implements most provider or provider factory traits.
+#[derive(Debug)]
+pub struct ShareableDatabase<DB> {
+    /// Database
+    db: DB,
+    /// Chain spec
+    chain_spec: Arc<ChainSpec>,
+}
+
+impl<DB: Database> ShareableDatabase<DB> {
+    /// Returns a provider with a created `DbTx` inside, which allows fetching data from the
+    /// database using different types of providers. Example: [`HeaderProvider`]
+    /// [`BlockHashProvider`]
+    pub fn provider(&self) -> Result<ProviderRO<'_, DB>> {
+        Ok(Provider::new(self.db.tx()?, self.chain_spec.clone()))
+    }
+
+    /// Returns a provider with a created `DbTxMut` inside, which allows fetching and updating
+    /// data from the database using different types of providers. Example: [`HeaderProvider`]
+    /// [`BlockHashProvider`]
+    pub fn provider_rw(&self) -> Result<ProviderRW<'_, DB>> {
+        Ok(Provider::new_rw(self.db.tx_mut()?, self.chain_spec.clone()))
+    }
```

Comment:
This is great - I think the main thing for us to consider is if we can pool these transaction openings somehow to juice extra performance when the RPC hammers the DB. Otherwise I like the flow of:

1. Instantiate DB
2. Get the right provider (RO vs RW) from it which opens a tx
3. Import the right provider traits, and go

Step 2 is optional and allows calling methods in the db which by default will open a new tx vs reusing each time